hail/src/main/scala/is/hail/linalg/BlockMatrix.scala

```diff
@@ -536,7 +378,60 @@ class BlockMatrix(val blocks: RDD[((Int, Int), BDM[Double])],
     
     filterBlocks(gp.rectanglesBlocks(rectangles))
   }
-  
+
+  def exportRectangles(
+    hc: HailContext,
+    output: String,
+    rectangles: Array[Array[Long]],
+    delimiter: String,
+    binary: Boolean) {
+
+    val writeRectangleBinary = (uos: OutputStream, dm: BDM[Double]) => {
+      val os = new DoubleOutputBuffer(uos, RichArray.defaultBufSize)
+      os.writeDoubles(dm.t.toArray)
+      os.close()
+    }
+
+    val writeRectangleText = (uos: OutputStream, dm: BDM[Double]) => {
+      val data = dm.t.toArray
+      val nCols = dm.cols
+      val os = new OutputStreamWriter(uos)
+      val sb = new StringBuilder(blockSize << 2)
+
+      var k = 0
+      while (k < data.length - 1) {
+        sb.append(data(k))
+        if (k % nCols == nCols - 1) {
+          sb.append("\n")
+        } else {
+          sb.append(delimiter)
+        }
+        k += 1
+      }
+      sb.append(data.last).append("\n")
+
+      os.write(sb.result())
+      os.close()
+    }
+
+    val writeRectangle = if (binary) writeRectangleBinary else writeRectangleText
+
+    val dRect = digitsNeeded(rectangles.length)
+    val sHadoopBc = hc.hadoopConfBc
+    BlockMatrixRectanglesRDD(rectangles, bm = this).foreach({ case (index, rectData) =>
+      val r = rectangles(index)
+      val paddedIndex = StringUtils.leftPad(index.toString, dRect, "0")
+      val outputFile = output + "/rect-" + paddedIndex + "_" + r.mkString("-")
+
+      if (rectData.size > 0) {
+        sHadoopBc.value.value.writeFile(outputFile) { uos =>
+          writeRectangle(uos, rectData)
+          uos.close()
```

Comment:
write file handles the close for you

Reply:
this can be `sHadoopBc.value.value.writeFile(outputFile)(writeRectangle(_, rectData))`

Reply:
That's awesome